File: server/app/schemas/analysis.py
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from datetime import datetime
from enum import Enum

class TimeRange(str, Enum):
    WEEK = "week"
    MONTH = "month"
    CUSTOM = "custom"

class ExportFormat(str, Enum):
    PDF = "pdf"
    JSON = "json"

class ContentInsight(BaseModel):
    category: str
    count: int
    trend: str  # "up", "down", "stable"
    keywords: List[str]
    percentage: float

class TimeInsight(BaseModel):
    completion_rate: float
    overdue_rate: float
    avg_completion_time_hours: float
    peak_working_hours: List[int]  # [9, 10, 11, 14, 15]
    trend: str  # "improving", "declining", "stable"

class ProductivityInsight(BaseModel):
    tasks_per_day: float
    completion_trend: str  # "increasing", "decreasing", "stable"
    peak_days: List[str]  # ["Monday", "Tuesday"]
    efficiency_score: float  # 0-100

class ComparisonData(BaseModel):
    current_period: Dict[str, Any]
    previous_period: Dict[str, Any]
    change_percentage: float
    improvement_areas: List[str]
    declining_areas: List[str]

class Recommendation(BaseModel):
    title: str
    description: str
    priority: str  # "high", "medium", "low"
    category: str  # "time_management", "productivity", "content_organization"

class AnalysisRequest(BaseModel):
    time_range: TimeRange = TimeRange.WEEK
    compare_with_previous: bool = True
    custom_start_date: Optional[datetime] = None
    custom_end_date: Optional[datetime] = None

class AnalysisResponse(BaseModel):
    user_id: str
    time_range: TimeRange
    generated_at: datetime
    content: ContentInsight
    time: TimeInsight
    productivity: ProductivityInsight
    comparison: Optional[ComparisonData] = None
    recommendations: List[Recommendation]
    confidence_score: float  # 0-100
    total_todos_analyzed: int

class ExportRequest(BaseModel):
    format: ExportFormat
    include_charts: bool = True
    time_range: TimeRange = TimeRange.WEEK

class ExportResponse(BaseModel):
    download_url: Optional[str] = None
    data: Optional[Dict[str, Any]] = None
    filename: str
    generated_at: datetime
    file_size_bytes: Optional[int] = None

class AnalysisHistoryItem(BaseModel):
    id: str
    time_range: TimeRange
    generated_at: datetime
    confidence_score: float
    total_todos: int
    key_insights: List[str]

class AnalysisHistoryResponse(BaseModel):
    items: List[AnalysisHistoryItem]
    total: int

File: client/src/types/analysis.ts
export type TimeRange = 'week' | 'month' | 'custom';

export type ExportFormat = 'pdf' | 'json';

export interface ContentInsight {
  category: string;
  count: number;
  trend: 'up' | 'down' | 'stable';
  keywords: string[];
  percentage: number;
}

export interface TimeInsight {
  completion_rate: number;
  overdue_rate: number;
  avg_completion_time_hours: number;
  peak_working_hours: number[];
  trend: 'improving' | 'declining' | 'stable';
}

export interface ProductivityInsight {
  tasks_per_day: number;
  completion_trend: 'increasing' | 'decreasing' | 'stable';
  peak_days: string[];
  efficiency_score: number;
}

export interface ComparisonData {
  current_period: Record<string, any>;
  previous_period: Record<string, any>;
  change_percentage: number;
  improvement_areas: string[];
  declining_areas: string[];
}

export interface Recommendation {
  title: string;
  description: string;
  priority: 'high' | 'medium' | 'low';
  category: 'time_management' | 'productivity' | 'content_organization';
}

export interface AnalysisRequest {
  time_range: TimeRange;
  compare_with_previous: boolean;
  custom_start_date?: string;
  custom_end_date?: string;
}

export interface AnalysisResponse {
  user_id: string;
  time_range: TimeRange;
  generated_at: string;
  content: ContentInsight;
  time: TimeInsight;
  productivity: ProductivityInsight;
  comparison?: ComparisonData;
  recommendations: Recommendation[];
  confidence_score: number;
  total_todos_analyzed: number;
}

export interface ExportRequest {
  format: ExportFormat;
  include_charts: boolean;
  time_range: TimeRange;
}

export interface ExportResponse {
  download_url?: string;
  data?: Record<string, any>;
  filename: string;
  generated_at: string;
  file_size_bytes?: number;
}

export interface AnalysisHistoryItem {
  id: string;
  time_range: TimeRange;
  generated_at: string;
  confidence_score: number;
  total_todos: number;
  key_insights: string[];
}

export interface AnalysisHistoryResponse {
  items: AnalysisHistoryItem[];
  total: number;
}

export interface AnalysisState {
  loading: boolean;
  error: string | null;
  analysisData: AnalysisResponse | null;
  isGenerating: boolean;
  isExporting: boolean;
}
